core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/MemorySailStore.java

```diff
@@ -511,7 +511,10 @@ public synchronized void deprecate(Statement statement) throws SailException {
 			acquireExclusiveTransactionLock();
 			requireCleanup = true;
 			if (statement instanceof MemStatement) {
-				((MemStatement) statement).setTillSnapshot(nextSnapshot);
+				MemStatement toDeprecate = (MemStatement) statement;
+				if (toDeprecate.getTillSnapshot() > nextSnapshot && toDeprecate.isExplicit() == explicit) {
+					toDeprecate.setTillSnapshot(nextSnapshot);
+				}
```

Comment:
This fixes a bug I believe I've introduced previously. Not a problem until now though.

Reply:
How is this fix related to the new SimpleUpgradeableModel class?

Reply:
I've made some more changes here.

The concept of directly manipulating the statement was introduced with the ElasticsearchStore. Gives a decent performance improvement since it allows us to skip retrieving a statement we already have. Unfortunately it's common that we try to deprecate a statement that can't be deprecated (not in our snapshot or explicit/inferred). That's usually handled when retrieving the statement again, however since we are working on it directly we need to apply the same checks.

Reply:
I've not copied this line from here: https://github.com/eclipse/rdf4j/blob/d2a301dabcf9c051b019e3434327c15c29fc321e/core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/model/MemStatementIterator.java#L142

Reply:
Btw. The reason this wasn't caught earlier was because the statements was never instanceof MemStatement directly until now. The LinkedHashModel wraps it's statement in a LinkedHashModelStatement (or something like that)